로또 생성기 작성

shinsanghooon opened the pull request:
로또 생성기 소스 코드 및 테스트 코드 작성

shinsanghooon committed:
로또 생성기 작성

kyucumber commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+/.gradle/
+/build/
+/build/classes/java/main/
+/build/classes/java/test/
```
POSIX 표준에 의하면 파일의 끝은 항상 newline(개행)으로 끝나야 합니다.

어떤 IDE를 사용하시나요? 파일의 끝에 newline을 붙이도록 IDE 설정을 해주시면 좋을 것 같아요.
아니면 editorconfig라는 설정에 아래와 같은 옵션을 추가해주실수도 있습니다.

```
insert_final_newline = true
```

editorconfig에는 indent(4칸으로 할지, 2칸으로 할지) 등등 규칙들을 정의할 수 있어요.
newline이든 editorconfig든 별 거 아닌 것들이지만 같이 협업할 때 이와 같은 부분은 생산성에 큰 영향을 끼칠 수 있으니 참고해주세요~

kyucumber commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,44 @@
+package lotto;
```
아마 알고 계실 것 같지만 일반적으로 패키지 네이밍을 어떻게 하는지 아래 문서를 참고해보셔도 좋을 것 같습니다.

https://programmers.tistory.com/entry/%ED%8C%A8%ED%82%A4%EC%A7%80%EC%9D%B4%EB%A6%84-%EB%A7%8C%EB%93%A4%EA%B8%B0-Package-Naming

kyucumber commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.util.*;
+
+public class Lotto {
```
Lotto라는 클래스의 이름만 보았을 때 Lotto와 관련된 도메인 정보를 담고있는 클래스라고 생각하지 않을까요?

로또의 정보와 특성(6개의 숫자를 가짐, 중복되지 않음, 1 - 46까지의 숫자를 가짐)을 표현할 수 있는 클래스를 만들고 해당 클래스에 로또와 관련된 기능들을 메소드로 정의해보시면서 객체지향과 관련된 연습을 해보시면 좋을 것 같아요.

아래와 같은 형태가 될 수 있겠네요.

```java
class Lotto {
    private final List<Integer> numbers;
}
```

아니면 Integer는 음수가 들어가거나 할 수도 있으니 아래와 같이 래핑된 클래스를 활용하실 수도 있습니다.

```java
class Lotto {
    private final List<PositiveNumber> numbers;
}
```

그리고 일급 컬렉션이라는 것도 자주 사용하는데 아래 내용을 참고해보시면 좋을 것 같아요.

https://jojoldu.tistory.com/412

kyucumber commented on src/main/java/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,45 @@
+        Set<Integer>[] lotto = new Set[generate_cnt];
+
+        for (int i = 0; i < generate_cnt; i++) {
+            lotto[i] = generateLottoOneLine();
```
Set의 배열을 리턴하신 이유가 `Set<Set<Integer>>`가 가독성이 좋지 않기 때문일까요?

일반적인 사용 예가 아니라 협업하는 개발자들에게 혼동을 줄 수 있을 것 같아요.

만약 가독성과 관련된 문제 때문이라면 일급 컬렉션 사용을 고려해보시는게 좋을 것 같습니다.

https://jojoldu.tistory.com/412

kyucumber commented on src/main/java/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,45 @@
+            lotto[i] = generateLottoOneLine();
+        }
+
+        return lotto;
```
이 부분 Stream.generate와 limit 등을 이용해 불변으로 처리해보세요.

```
Stream.generate(--)
                    .limit(generate_cnt)
                    .collect(Collectors.toList());
```

kyucumber commented on src/main/java/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,45 @@
+            throw new IllegalArgumentException("구입 금액은 1000원 이상이어야 합니다.");
+        }
+
+        int generate_cnt = money / 1000;
```
자바에서는 SnakeCase가 아닌 CamelCase를 사용합니다.

사소한 부분이지만 언어에 대한 지식이 없다고 비춰질 수 있으므로 컨벤션을 잘 지켜주세요.

kyucumber commented on src/main/java/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,45 @@
+     * @return 생성된 로또 리스트
+     */
+    public Set<Integer>[] getLotto(int money) {
+        if (money < 1000) {
```
이런 1000과 같은 상수는 도메인 전문가가 아니면 무엇을 의미하는지 알기 어렵기 때문에 상수로 추출해도 좋을 것 같아요.

kyucumber commented on src/main/java/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,45 @@
+     * @param money 구매 금액
+     * @return 생성된 로또 리스트
+     */
+    public Set<Integer>[] getLotto(int money) {
```
메소드 명은 Lotto라 로또 한장을 리턴할 것 같지만 실제로는 Lotto의 리스트를 리턴하고 있어요.

`getLottos`는 어떨까요?

kyucumber commented on src/main/java/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,45 @@
+        while (!(set.size() == numberOfOneLine)) {
+            set.add((int) (Math.random() * 45) + 1);
+        }
+        return set;
```
Stream 사용을 좀 더 연습해보시면 좋을 것 같아요.

사실 성능은 이렇게 하는게 더 좋겠지만 아래 문서를 참고해 스트림으로 변경해보세요.

https://stackoverflow.com/questions/20058366/shuffle-a-list-of-integers-with-java-8-streams-api

kyucumber commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,44 @@
+        System.out.println("구매 금액을 입력해 주세요.");
+        
+        Scanner sc = new Scanner(System.in);
+        int money = sc.nextInt();
```
이 부분도 입력하는 부분을 별도로 분리해볼 수 있겠네요.

kyucumber commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,44 @@
+        LottoGenerator lottoGenerator = new LottoGenerator(NUMBER_OF_ONE_LINE);
+        Set<Integer>[] myLotto = lottoGenerator.getLotto(money);
+
+        System.out.println(myLotto.length + "개를 구매했습니다.");
+
+        for (Set<Integer> lotto : myLotto) {
+            System.out.println(lotto.toString());
+        }
+
+        System.out.println();
+        System.out.println("지난 주 당첨 번호를 입력해주세요.");
+
+        Set<Integer> winningNumberSet = new HashSet<>();
+        for (int i = 0; i < NUMBER_OF_ONE_LINE; i++) {
+            String next = sc.next().split(",")[0];
+            winningNumberSet.add(Integer.parseInt(next));
+        }
+
+        LottoWinningChecker winningChecker = new LottoWinningChecker();
+        Map<LottoWinningEnum, Integer> result = winningChecker.execute(myLotto, winningNumberSet);
+
+        System.out.println();
+        System.out.println("당첨 통계");
+        System.out.println("-----------------");
```
여러 클래스들을 잘 분리해주셨지만 main 로직 내에 코드가 많은 것 같아요.

웹이 아니라 조금 비유가 잘못되었을 수 있는데 컨트롤러에 서비스 로직들이 많이 섞여있는 것과 같은 형태라고 표현할수도 있겠네요.

모든 print 부분을 출력을 위한 view 패키지 등으로 잘 분리해서 main 로직의 코드를 최대한 줄여보시면 좋겠습니다.

kyucumber commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,44 @@
+        System.out.println();
+        System.out.println("지난 주 당첨 번호를 입력해주세요.");
+
+        Set<Integer> winningNumberSet = new HashSet<>();
+        for (int i = 0; i < NUMBER_OF_ONE_LINE; i++) {
+            String next = sc.next().split(",")[0];
+            winningNumberSet.add(Integer.parseInt(next));
+        }
```
Input과 관련된 부분이고 MVC로 보면 view에 해당하는 부분이 될 수 있을 것 같아요. Scanner 부분, 구매 금액 입력 / 당첨번호 입력 등을 별도의 클래스로 추출해보아도 좋을 것 같습니다.

kyucumber commented on src/main/java/lotto/LottoWinningChecker.java:
```diff
@@ -0,0 +1,65 @@
+
+            int result = myLottoOneLine.size();
+            if (result >= MINIMUM_NUMBER_OF_WINNING) {
+                addResultToWinningMap(lottoWinningMap, result);
+            }
+        }
+
+        return lottoWinningMap;
+    }
+
+
+    /**
+     * 로또 당첨 횟수를 순위마다 0으로 초기화 한다
+     * @param lottoWinningMap 로또 당첨 현황 Map
+     */
+    private void initializeMap(Map<LottoWinningEnum, Integer> lottoWinningMap) {
+        lottoWinningMap.put(LottoWinningEnum.THREE, 0);
+        lottoWinningMap.put(LottoWinningEnum.FOUR, 0);
+        lottoWinningMap.put(LottoWinningEnum.FIVE, 0);
+        lottoWinningMap.put(LottoWinningEnum.SIX, 0);
+    }
+
+
+    /**
+     * 당첨 순위에 따라 카운트를 1개씩 증가시킨다.
+     * @param lottoWinningMap 로또 당첨 현황 Map
+     * @param result 로또 한 줄에 당첨된 번호의 수
+     */
+    private void addResultToWinningMap(Map<LottoWinningEnum, Integer> lottoWinningMap, int result) {
+        if (result == 3) {
+            lottoWinningMap.put(LottoWinningEnum.THREE, lottoWinningMap.get(LottoWinningEnum.THREE) + 1);
+        } else if (result == 4) {
+            lottoWinningMap.put(LottoWinningEnum.FOUR, lottoWinningMap.get(LottoWinningEnum.FOUR) + 1);
+        } else if (result == 5) {
+            lottoWinningMap.put(LottoWinningEnum.FIVE, lottoWinningMap.get(LottoWinningEnum.FIVE) + 1);
+        } else {
+            // result = 6
+            lottoWinningMap.put(LottoWinningEnum.SIX, lottoWinningMap.get(LottoWinningEnum.SIX) + 1);
+        }
+    }
```
결과를 Map에 잘 담아주셨네요 👍 

여기서 Stream을 좀 더 심화적으로 학습하고 싶으시면 아래와 같은 함수들을 찾아보시고 적용해보셔도 좋을 것 같네요.

```java
lottoWinningEnums.stream()
                        .collect(groupingBy(
                                Function.identity(),
                                () -> new EnumMap<>(LottoWinningEnum.class),
                                summingInt(-)))
```

kyucumber replied:
Stream이 알고리즘 풀거나 할 땐 성능이 좋지 않지만 잘 쓰면 이런 복잡한 로직을 체이닝 된 메소드 하나로 처리할 수 있습니다.

지금은 현업에서 대부분 자바 8, 거의 11을 많이 사용하니 Stream은 꼭 연습해보시는게 좋습니다.

kyucumber commented on src/test/java/lotto/LottoWinningCheckerTest.java:
```diff
@@ -0,0 +1,136 @@
+    @DisplayName("6개의 숫자를 1개만 맞춘 경우, 당첨 로또 개수를 반환한다.")
+    @Test
+    void sixMatchesLineCount() {
+        // given
+        Set<Integer>[] myLotto = new Set[1];
+        HashSet<Integer> lotto = new HashSet<>(Arrays.asList(1, 2, 3, 4, 5, 6));
+        myLotto[0] = lotto;
+
+        // when
+        Map<LottoWinningEnum, Integer> result =
+                winningChecker.execute(myLotto, winningNumberSet);
+
+        // then
+        assertThat(result.get(LottoWinningEnum.THREE)).isEqualTo(0);
+        assertThat(result.get(LottoWinningEnum.FOUR)).isEqualTo(0);
+        assertThat(result.get(LottoWinningEnum.FIVE)).isEqualTo(0);
+        assertThat(result.get(LottoWinningEnum.SIX)).isEqualTo(1);
+    }
+
+
+    @DisplayName("3개의 숫자를 1개, 4개를 1개 맞춘 경우 당첨 로또 개수를 반환한다.")
+    @Test
+    void multipleMatchesLine() {
+        // given
+        Set<Integer>[] myLotto = new Set[2];
+        HashSet<Integer> lotto1 = new HashSet<>(Arrays.asList(1, 2, 3, 14, 15, 16));
+        HashSet<Integer> lotto2 = new HashSet<>(Arrays.asList(1, 2, 3, 4, 15, 16));
+        myLotto[0] = lotto1;
+        myLotto[1] = lotto2;
+
+        // when
+        Map<LottoWinningEnum, Integer> result =
+                winningChecker.execute(myLotto, winningNumberSet);
+
+        // then
+        assertThat(result.get(LottoWinningEnum.THREE)).isEqualTo(1);
+        assertThat(result.get(LottoWinningEnum.FOUR)).isEqualTo(1);
+        assertThat(result.get(LottoWinningEnum.FIVE)).isEqualTo(0);
+        assertThat(result.get(LottoWinningEnum.SIX)).isEqualTo(0);
+    }
```
이런 경우라면 ParameterizedTest, methodSource 등을 활용하면 테스트 메소드 하나로도 테스트 할 수 있을 것 같아요.

관련해서 한번 학습하고 적용해보세요.

kyucumber commented on src/main/java/lotto/LottoResultReviewer.java:
```diff
@@ -0,0 +1,30 @@
+     * 로또 결과를 프린트한다.
+     * @param lottoWinningMap 로또 결과가 담긴 Map
+     */
+    static public void printLottoResult(Map<LottoWinningEnum, Integer> lottoWinningMap) {
+        for (LottoWinningEnum lottoAwardEnum : lottoWinningMap.keySet()) {
+            int count = lottoWinningMap.get(lottoAwardEnum);
+            System.out.println(lottoAwardEnum.getDesc() + " (" + lottoAwardEnum.getPrizeMoney() + "원) - " + count + "개");
+        }
+    }
+
+    /**
+     * 수익률을 프린트한다.
+     * @param money 로또 구매에 사용한 금액
+     * @param lottoWinningMap 로또 결과가 담긴 Map
+     */
+    static public void calculateReturnRate(int money, Map<LottoWinningEnum, Integer> lottoWinningMap) {
+        int sum = 0;
+        for (LottoWinningEnum lottoAwardEnum : lottoWinningMap.keySet()) {
+            sum += lottoAwardEnum.getPrizeMoney() * lottoWinningMap.get(lottoAwardEnum);
+        }
+        System.out.println("총 수익률은 " + sum * 1.0 / money * 1.0 * 100 + "% 입니다.");
+    }
```
여기에 출력 부분이 섞여있네요.

개인적으로는 출력 부분은 비즈니스 로직에서 제거하고 여기서는 수익률을 반환하는게 좋다고 생각해요.

지금은 테스트가 어렵지만 수익률을 반환하는 로직으로 변경된다면 테스트도 수행할 수 있겠죠

kyucumber commented on src/main/java/lotto/LottoWinningChecker.java:
```diff
@@ -0,0 +1,65 @@
+     * @param result 로또 한 줄에 당첨된 번호의 수
+     */
+    private void addResultToWinningMap(Map<LottoWinningEnum, Integer> lottoWinningMap, int result) {
+        if (result == 3) {
+            lottoWinningMap.put(LottoWinningEnum.THREE, lottoWinningMap.get(LottoWinningEnum.THREE) + 1);
```
result가 3일 때 그에 해당하는 enum을 가져오는 메소드가 enum에 있다면

이런 복잡한 if, else가 여러개 중첩될 필요가 없지 않을까요?

예를 들면 아래와 같은 형태가 될 수 있겠네요.

```java
LottoWinningEnum a = LottoWinningEnum.getBy(result)
lottoWinningMap.put(a, lottoWinningMap.get(a) + 1);
```

shinsanghooon committed:
패키지 네이밍 변경

- lotto -> com.ssh.lotto로 변경

shinsanghooon committed:
.editorconfig 설정 파일 추가

- charset
- end of line style
- identatioin style
- etc ..

shinsanghooon committed:
Lotto 파일명 LottoApp으로 변경

- 기존 Lotto 클래스가 메인 클래스 역할을 하였지만 Lotto와 관련된 도메인
클래스 생성을 위해 이름을 변경

shinsanghooon committed:
Lotto 일급 컬렉션 적용

- Lotto 일급 컬렉션 적용
- ArrayList<Lotto>로 표현이 간긍해짐에 따라 가독성 향상

shinsanghooon committed:
당첨 결과 출력 분리

- 당첨 결과 출력 부분 로직 분리

shinsanghooon committed:
입출력 부분 로직 분리

- 구매 금액 입력 분리
- 구매한 로또 리스트 출력하는 분리
- 로또 당첨 번호 입력 분리
- 로또 결과 출력 분리
- 수익률 계산 로직 분리

shinsanghooon committed:
getLotto -> getLottos 로 메서드 이름 변경

- getLotto는 한 장의 로또를 리턴하는 느낌이 강한데, 실제로는 여러 장의
로또 리스트를 반환하기 때문에 변경함

shinsanghooon committed:
로또 한 장 구입에 필요한 금액을 상수로 추출

- 코드에서 1000 이라는 숫자를 사용하는 것보다 상수로 추출해서 1000이
의미하는 바를 나타내고자 수정

shinsanghooon committed:
로또 등수 enum을 가져오는 메서드 생성

- result에 따라 해당하는 로또 등수 enum을 가져오는 메서드 생성으로
if-else 문에서의 중첩 제거

shinsanghooon committed:
테스트 코드에 @ParamterizedTest, @MethodSource 적용

- @ParamterizedTest, @MethodSource 적용을 통해 테스트 메서드 중복 제거

shinsanghooon committed:
수익률 계산 테스트 코드 작성

- 수익률 계산 테스트 코드 작성

kyucumber commented on src/main/java/com/ssh/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+        return new HashSet<Integer>(lotto);
+    }
+
+    public int getSize() {
+        return lotto.size();
+    }
```
이건 테스트에서만 쓰이는 것 같네요.

테스트를 위한 메소드를 작성하는건 좋은 구현이 아니라고 생각되어요.

kyucumber commented on src/main/java/com/ssh/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    // 불변성을 유지 하기 위해 새로운 set를 만들어서 리턴하는게 맞나 ...
+    public Set<Integer> getLotto() {
+        return new HashSet<Integer>(lotto);
+    }
```
Collections.unmodifiableSet을 사용하시면 어떨까요?

kyucumber commented on src/main/java/com/ssh/lotto/LottoGenerator.java:
```diff
@@ -0,0 +1,50 @@
+     * @return 생성된 로또 1줄
+     */
+    private static Lotto generateLottoOneLine() {
+        Set<Integer> set = new TreeSet<>();
+        while (!(set.size() == numberOfOneLine)) {
+            set.add((int) (Math.random() * 45) + 1);
+        }
```
여기도 Collections.shuffle과 subLit등을 활용해 스트림을 사용해보세요.

kyucumber commented on src/main/java/com/ssh/lotto/view/WinningNumberInsertionView.java:
```diff
@@ -0,0 +1,25 @@
+
+        System.out.println("지난 주 당첨 번호를 입력해주세요.");
+
+        Set<Integer> winningNumberSet = new HashSet<>();
+        for (int i = 0; i < numberOfLine; i++) {
+            String next = sc.next().split(",")[0];
+            winningNumberSet.add(Integer.parseInt(next));
+        }
+
+        System.out.println();
+
+        return winningNumberSet;
```
당첨 번호인 6개의 숫자 또한 로또로 볼 수 있지 않을까요?

Set<Integer>로 반환하는 경우 기존 Lotto에서 검증되던 부분들이 누락될 것 같아요.

Lotto의 validation을 활용하고 추가로 필요한 기능(보너스 번호)을 충족하기 위해 Composition을 활용해볼수도 있을 것 같습니다.

kyucumber commented on src/main/java/com/ssh/lotto/LottoWinningEnum.java:
```diff
@@ -0,0 +1,38 @@
+        return desc;
+    }
+
+    public static LottoWinningEnum getResult(int result) {
+        for (LottoWinningEnum e : values()) {
+            if (e.correctCnt == result) {
+                return e;
+            }
+        }
+        return null;
```
null 보다는 java8에서 제공되는 Optional을 활용해보시면 좋을 것 같아요.

그리고 당첨되지 않은 경우를 표현하는 Enum이 필요할지에 대한 부분도 고민해보시면 좋을 것 같네요.

kyucumber commented on src/main/java/com/ssh/lotto/LottoWinningChecker.java:
```diff
@@ -0,0 +1,56 @@
+        Map<LottoWinningEnum, Integer> lottoWinningMap = new TreeMap<>();
+        initializeMap(lottoWinningMap);
+
+        for (Lotto myLottoOneLine : myLotto) {
+            Set<Integer> lotto = myLottoOneLine.getLotto();
+            lotto.retainAll(winningNumberSet);
+
+            int result = lotto.size();
+            if (result >= MINIMUM_NUMBER_OF_WINNING) {
+                addResultToWinningMap(lottoWinningMap, result);
+            }
+        }
+
+        return lottoWinningMap;
```
이전에 코멘트 남기긴 했는데 아래와 같은 형태로 스트림을 활용하는 연습을 해보시면 좋을 것 같습니다.

```java
lottoWinningEnums.stream()
                        .collect(groupingBy(
                                Function.identity(),
                                () -> new EnumMap<>(LottoWinningEnum.class),
                                summingInt(-)))
```

kyucumber commented on src/main/java/com/ssh/lotto/view/LottoResultPrintView.java:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    static public void calculateReturnRate(int returnRate) {
```
여기서는 void 리턴 타입으로 내부에서 출력만을 하고 있는데 calculate라는 동사가 가지는 의미 때문에 어떤 결과를 반환할 것으로 착각할 수 있을 것 같아요.

다른쪽과 동일하게 print~~로 하는건 어떨까요?
